third_party/skia/platform_view_rtree.cc

```diff
@@ -0,0 +1,244 @@
+            currentBranch++;
+        }
+    }
+    return nodes + CountNodes(nodes);
+}
+
+PlatformViewRTree::Branch PlatformViewRTree::bulkLoad(std::vector<Branch>* branches, int level) {
+    if (branches->size() == 1) {  // Only one branch.  It will be the root.
+        return (*branches)[0];
+    }
+
+    // We might sort our branches here, but we expect Blink gives us a reasonable
+    // x,y order. Skipping a call to sort (in Y) here resulted in a 17% win for
+    // recording with negligible difference in playback speed.
+    int numBranches = (int)branches->size() / kMaxChildren;
+    int remainder = (int)branches->size() % kMaxChildren;
+    int newBranches = 0;
+
+    if (remainder > 0) {
+        ++numBranches;
+        // If the remainder isn't enough to fill a node, we'll add fewer nodes to
+        // other branches.
+        if (remainder >= kMinChildren) {
+            remainder = 0;
+        } else {
+            remainder = kMinChildren - remainder;
+        }
+    }
+
+    int currentBranch = 0;
+    while (currentBranch < (int)branches->size()) {
+        int incrementBy = kMaxChildren;
+        if (remainder != 0) {
+            // if need be, omit some nodes to make up for remainder
+            if (remainder <= kMaxChildren - kMinChildren) {
+                incrementBy -= remainder;
+                remainder = 0;
+            } else {
+                incrementBy = kMinChildren;
+                remainder -= kMaxChildren - kMinChildren;
+            }
+        }
+        Node* n = allocateNodeAtLevel(level);
+        n->fNumChildren = 1;
+        n->fChildren[0] = (*branches)[currentBranch];
+        Branch b;
+        b.fBounds = (*branches)[currentBranch].fBounds;
+        b.fSubtree = n;
+        ++currentBranch;
+        for (int k = 1; k < incrementBy && currentBranch < (int)branches->size(); ++k) {
+            b.fBounds.join((*branches)[currentBranch].fBounds);
+            n->fChildren[k] = (*branches)[currentBranch];
+            ++n->fNumChildren;
+            ++currentBranch;
+        }
+        (*branches)[newBranches] = b;
+        ++newBranches;
+    }
+    branches->resize(newBranches);
+    return this->bulkLoad(branches, level + 1);
+}
+
+void PlatformViewRTree::search(const SkRect& query, std::vector<int>* results) const {
+    if (fCount > 0 && SkRect::Intersects(fRoot.fBounds, query)) {
+        this->search(fRoot.fSubtree, query, results);
+    }
+}
+
+void PlatformViewRTree::search(Node* node, const SkRect& query, std::vector<int>* results) const {
+    for (int i = 0; i < node->fNumChildren; ++i) {
+        if (!SkRect::Intersects(node->fChildren[i].fBounds, query)) {
+            continue;
+        }
+        if (0 != node->fLevel) {
+            this->search(node->fChildren[i].fSubtree, query, results);
+            continue;
+        }
+        results->push_back(node->fChildren[i].fOpIndex);
+    }
+}
+
+std::list<SkRect> PlatformViewRTree::searchRects(const SkRect& query) const {
```

Comment:
Would it be preferable to return a pointer and avoid copying the list on return?

Reply:
@chinmaygarde suggested to don't return a pointer. Are we concerned about the cost of copying the list? We are already copying the SkRects from the tree since we are mutating them.